lib/prelude.dx

```diff
@@ -40,105 +40,105 @@ def FToI (x:Float) : Int = internalCast _ x
 def I64ToRawPtr (x:Int64 ) : RawPtr = internalCast _ x
 def RawPtrToI64 (x:RawPtr) : Int64  = internalCast _ x
 
-interface Add a:Type where
+interface Add a
   add : a -> a -> a
   sub : a -> a -> a
   zero : a
 
 def (+) [Add a] : a -> a -> a = add
 def (-) [Add a] : a -> a -> a = sub
 
-instance float64Add : Add Float64 where
-  add = \x:Float64 y:Float64. %fadd x y
-  sub = \x:Float64 y:Float64. %fsub x y
+instance Add Float64
+  add = \x y. %fadd x y
+  sub = \x y. %fsub x y
   zero = FToF64 0.0
 
-instance float32Add : Add Float32 where
-  add = \x:Float32 y:Float32. %fadd x y
-  sub = \x:Float32 y:Float32. %fsub x y
+instance Add Float32
+  add = \x y. %fadd x y
+  sub = \x y. %fsub x y
   zero = FToF32 0.0
 
-instance int64Add : Add Int64 where
-  add = \x:Int64 y:Int64. %iadd x y
-  sub = \x:Int64 y:Int64. %isub x y
+instance Add Int64
+  add = \x y. %iadd x y
+  sub = \x y. %isub x y
   zero = IToI64 0
 
-instance int32Add : Add Int32 where
-  add = \x:Int32 y:Int32. %iadd x y
-  sub = \x:Int32 y:Int32. %isub x y
+instance Add Int32
+  add = \x y. %iadd x y
+  sub = \x y. %isub x y
   zero = IToI32 0
 
-instance word8Add : Add Word8 where
-  add = \x:Word8 y:Word8. %iadd x y
-  sub = \x:Word8 y:Word8. %isub x y
+instance Add Word8
+  add = \x y. %iadd x y
```

Comment:
Did this patch make inference powerful enough, or was it something else?

Reply:
Yes, this patch made it powerful enough, by bringing the types declared in the interface decl to the instance decls.